File: sdks/go/batch.go
package voidcontrol

import "time"

type BatchClient struct {
	client    *Client
	routeBase string
}

type BatchRunsClient struct {
	client    *Client
	routeBase string
}

func (client *BatchClient) Run(spec map[string]any) (*BatchRunResult, error) {
	var response BatchRunResult
	if err := client.client.postJSON(client.routeBase+"/run", spec, &response); err != nil {
		return nil, err
	}
	return &response, nil
}

func (client *BatchRunsClient) Get(runID string) (*BatchRunDetail, error) {
	var response BatchRunDetail
	if err := client.client.getJSON(client.routeBase+"-runs/"+runID, &response); err != nil {
		return nil, err
	}
	return &response, nil
}

func (client *BatchRunsClient) Wait(runID string) (*BatchRunDetail, error) {
	for {
		detail, err := client.Get(runID)
		if err != nil {
			return nil, err
		}
		if detail.Execution.Status == "Completed" || detail.Execution.Status == "Failed" || detail.Execution.Status == "Canceled" {
			return detail, nil
		}
		time.Sleep(10 * time.Millisecond)
	}
}

File: sdks/node/src/batch.js
const TERMINAL_STATUSES = new Set(["Completed", "Failed", "Canceled"]);

export class BatchClient {
  constructor(client, { routeBase }) {
    this._client = client;
    this._routeBase = routeBase;
  }

  async run(spec) {
    const payload = await this._client.postJson(`${this._routeBase}/run`, spec);
    return {
      kind: String(payload.kind),
      runId: String(payload.run_id),
      executionId: String(payload.execution_id),
      compiledPrimitive: String(payload.compiled_primitive),
      status: String(payload.status),
      goal: String(payload.goal)
    };
  }
}

export class BatchRunsClient {
  constructor(client, { routeBase }) {
    this._client = client;
    this._routeBase = routeBase;
  }

  async get(runId) {
    const payload = await this._client.getJson(`${this._routeBase}-runs/${runId}`);
    return {
      kind: String(payload.kind),
      runId: String(payload.run_id),
      execution: {
        executionId: String(payload.execution.execution_id),
        goal: String(payload.execution.goal),
        status: String(payload.execution.status)
      },
      progress: payload.progress ?? {},
      result: {
        bestCandidateId:
          payload.result?.best_candidate_id == null
            ? null
            : String(payload.result.best_candidate_id),
        completedIterations: Number(payload.result?.completed_iterations ?? 0),
        totalCandidateFailures: Number(payload.result?.total_candidate_failures ?? 0)
      },
      candidates: payload.candidates ?? []
    };
  }

  async wait(runId, { pollIntervalMs = 1000 } = {}) {
    while (true) {
      const detail = await this.get(runId);
      if (TERMINAL_STATUSES.has(detail.execution.status)) {
        return detail;
      }
      await new Promise((resolve) => setTimeout(resolve, pollIntervalMs));
    }
  }
}
